dartagnan/src/main/java/com/dat3m/dartagnan/expression/ExpressionPrinter.java

```diff
@@ -0,0 +1,98 @@
+package com.dat3m.dartagnan.expression;
+
+import com.dat3m.dartagnan.expression.floats.FloatSizeCast;
+import com.dat3m.dartagnan.expression.floats.IntToFloatCast;
+import com.dat3m.dartagnan.expression.integers.FloatToIntCast;
+import com.dat3m.dartagnan.expression.integers.IntSizeCast;
+import com.dat3m.dartagnan.expression.misc.ConstructExpr;
+import com.dat3m.dartagnan.expression.misc.ExtractExpr;
+import com.dat3m.dartagnan.expression.misc.GEPExpr;
+import com.dat3m.dartagnan.expression.misc.ITEExpr;
+import com.dat3m.dartagnan.program.Register;
+
+import java.util.stream.Collectors;
+
+public final class ExpressionPrinter implements ExpressionVisitor<String> {
+
+    private final boolean qualified;
+
+    private ExpressionKind kind;
+
+    public ExpressionPrinter(boolean qualified) {
+        this.qualified = qualified;
+    }
+
+    public String visit(Expression expr) {
+        final ExpressionKind kind = this.kind;
+        this.kind = expr.getKind();
+        final String inner = expr.accept(this);
+        final boolean direct = kind == null || kind == this.kind || expr.getOperands().isEmpty();
+        final String result = direct ? inner : "(" + inner + ")";
+        this.kind = kind;
+        return result;
```

Comment:
Can one of the variables (probably the local one) be renamed to something like `parentKind`?
I found this very hard to read because the two kinds that get compared have the same names.

Btw. for all non-associative operations, this code creates ambiguous outputs (e.g. for `-` and `\`).
It might not be too bad, since we rarely have those nested directly, but it deserves at least a `TOOD/FIXME` comment.